Razor runtime compilation produces errors if running on a shared runtime that's rolled forward

1) Create an image with 2.1.1 runtime
2) `dotnet new Mvc`. Update the project file to specify a version for the shared runtime. Specify an older runtime e.g. `<PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.0" />`
3) Run the app. Views would build and be served.
4) Edit a view while the app is running and refresh the browser.

The compilation produces an error:
```
qqbcsiev.hbu(50,67): error CS0433: The type 'UrlResolutionTagHelper' exists in both 'Microsoft.AspNetCore.Mvc.Razor, Version=2.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' and 'Microsoft.AspNetCore.Mvc.Razor, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
qqbcsiev.hbu(214,58): error CS0433: The type 'RazorInjectAttribute' exists in both 'Microsoft.AspNetCore.Mvc.Razor, Version=2.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' and 'Microsoft.AspNetCore.Mvc.Razor, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
```

--- a/src/Microsoft.AspNetCore.Mvc/MvcServiceCollectionExtensions.cs
+++ b/src/Microsoft.AspNetCore.Mvc/MvcServiceCollectionExtensions.cs
@@ -2,6 +2,8 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 using System.Reflection;
 using Microsoft.AspNetCore.Mvc;
@@ -59,15 +61,15 @@ namespace Microsoft.Extensions.DependencyInjection
         private static void AddDefaultFrameworkParts(ApplicationPartManager partManager)
         {
             var mvcTagHelpersAssembly = typeof(InputTagHelper).GetTypeInfo().Assembly;
-            if(!partManager.ApplicationParts.OfType<AssemblyPart>().Any(p => p.Assembly == mvcTagHelpersAssembly))
+            if (!partManager.ApplicationParts.OfType<AssemblyPart>().Any(p => p.Assembly == mvcTagHelpersAssembly))
             {
-                partManager.ApplicationParts.Add(new AssemblyPart(mvcTagHelpersAssembly));
+                partManager.ApplicationParts.Add(new FrameworkAssemblyPart(mvcTagHelpersAssembly));
             }
-            
+
             var mvcRazorAssembly = typeof(UrlResolutionTagHelper).GetTypeInfo().Assembly;
-            if(!partManager.ApplicationParts.OfType<AssemblyPart>().Any(p => p.Assembly == mvcRazorAssembly))
+            if (!partManager.ApplicationParts.OfType<AssemblyPart>().Any(p => p.Assembly == mvcRazorAssembly))
             {
-                partManager.ApplicationParts.Add(new AssemblyPart(mvcRazorAssembly));
+                partManager.ApplicationParts.Add(new FrameworkAssemblyPart(mvcRazorAssembly));
             }
         }
 
@@ -93,6 +95,17 @@ namespace Microsoft.Extensions.DependencyInjection
             builder.Services.Configure(setupAction);
 
             return builder;
+        }
+
+        [DebuggerDisplay("{Name}")]
+        private class FrameworkAssemblyPart : AssemblyPart, ICompilationReferencesProvider
+        {
+            public FrameworkAssemblyPart(Assembly assembly)
+                : base(assembly)
+            {
+            }
+
+            IEnumerable<string> ICompilationReferencesProvider.GetReferencePaths() => Enumerable.Empty<string>();
         }
     }
 }

--- a/test/WebSites/BasicWebSite/Controllers/HomeController.cs
+++ b/test/WebSites/BasicWebSite/Controllers/HomeController.cs
@@ -129,7 +129,7 @@ namespace BasicWebSite.Controllers
             // Ensures that the entry assembly part is marked correctly.
             var assemblyPartMetadata = applicationPartManager
                 .ApplicationParts
-                .Where(part => part.GetType() == typeof(AssemblyPart))
+                .OfType<AssemblyPart>()
                 .Select(part => part.Name)
                 .ToArray();
 

